### PR TITLE
better histogram for summary of discrete action

### PR DESCRIPTION
**Pull request by witwolf**
#43

### CODE REVIEW

File: alf/utils/summary_utils.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+from tensorboard.compat import tf2 as tf
+from tensorboard.plugins.histogram import metadata
+
+
+def histogram_discrete(name, data, buckets, step=None, description=None):
+    """histogram for discrete distributions
+
+    Args:
+        name (str): name for this summary
+        data (Tensor): A `Tensor` of any shape. Must be castable to `float64`
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: data should be integer?

File: alf/utils/summary_utils.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    with summary_scope(
+            name, 'histogram_summary', values=[data, buckets, step]) as (tag,
+                                                                         _):
+        with tf.name_scope('buckets'):
+            bucket_count = buckets
+            one_hots = tf.one_hot(
+                tf.reshape(data, shape=[-1]), depth=bucket_count)
+            bucket_counts = tf.cast(
+                tf.reduce_sum(input_tensor=one_hots, axis=0), tf.float64)
+            edge = tf.cast(tf.range(bucket_count), tf.float64)
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: should  action_spec.maximum and action_spec.minimum be taken into account for creating edge?

File: alf/drivers/policy_driver.py [hunk 3]
```diff
@@ -173,12 +173,14 @@ def predict(self, max_num_steps, time_step, policy_state):
...
         return time_step, policy_state
 
-    def _summary(self, loss_info, grads_and_vars):
+    def _summary(self, training_info, loss_info, grads_and_vars):
         if self._summarize_grads_and_vars:
             eager_utils.add_variables_summaries(grads_and_vars,
                                                 self._train_step_counter)
             eager_utils.add_gradients_summaries(grads_and_vars,
                                                 self._train_step_counter)
+            common.add_action_summaries(training_info.action,
```
Comment (emailweixu) [inline alf/drivers/policy_driver.py]: I didn't notice that  common.add_action_summaries() was removed in off_policy_driver PR.

File: alf/utils/common.py [hunk 4]
```diff
@@ -147,17 +148,23 @@ def add_action_summaries(actions, action_specs):
     """
     action_specs = tf.nest.flatten(action_specs)
     actions = tf.nest.flatten(actions)
+
     for i, (action, action_spec) in enumerate(zip(actions, action_specs)):
         if len(action_spec.shape) > 1:
             continue
-        if len(action_spec.shape) == 0:
-            action_dim = 1
+        if tensor_spec.is_discrete(action_spec):
```
Comment (emailweixu) [inline alf/utils/common.py]: line 155-158 should be in the loop of 165 in case it's multi-dimensional.

File: alf/utils/summary_utils.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from tensorboard.compat import tf2 as tf
+from tensorboard.plugins.histogram import metadata
+
+
+def histogram_discrete(name, data, buckets, step=None, description=None):
+    """histogram for discrete distributions
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: Histogram for discrete data

File: alf/utils/common.py (hunk 4 above)
Comment (witwolf) [reply alf/utils/common.py]: action_dim is 1 for discrete actions

File: alf/utils/summary_utils.py (hunk 2 above)
Comment (witwolf) [reply alf/utils/summary_utils.py]: action_spec.minimum is zero for discrete actions 

https://github.com/tensorflow/agents/blob/d7905cfc8b317900657113c34f51244ec5211b69/tf_agents/environments/gym_wrapper.py#L60-L80

File: alf/utils/summary_utils.py (hunk 1 above)
Comment (witwolf) [reply alf/utils/summary_utils.py]: updated

File: alf/utils/summary_utils.py (hunk 5 above)
Comment (witwolf) [reply alf/utils/summary_utils.py]: updated

**Pull request merged**